SYCL/SpecConstants/2020/handler-api.cpp

```diff
@@ -0,0 +1,199 @@
+
+    custom_type custom_type_ref;
+    if (!check_value(
+            custom_type_ref, cgh.get_specialization_constant<custom_type_id>(),
+            "custom_type specializaiton constant before setting any value"))
+      ++errors;
+
+    int new_int_value = 8;
+    double new_double_value = 3.0;
+    custom_type new_custom_type_value('b', 1.0, 12);
+    cgh.set_specialization_constant<int_id>(new_int_value);
+    cgh.set_specialization_constant<double_id>(new_double_value);
+    cgh.set_specialization_constant<custom_type_id>(new_custom_type_value);
+
+    if (!check_value(
+            new_int_value, cgh.get_specialization_constant<int_id>(),
+            "integer specializaiton constant after setting a new value"))
+      ++errors;
+
+    if (!check_value(
+            new_double_value, cgh.get_specialization_constant<double_id>(),
+            "double specializaiton constant after setting a new value"))
+      ++errors;
+
+    if (!check_value(
+            new_custom_type_value,
+            cgh.get_specialization_constant<custom_type_id>(),
+            "custom_type specializaiton constant after setting a new value"))
+      ++errors;
+
+    cgh.single_task<EmptyKernel>([=]() {});
+  });
+
+  return errors == 0;
+}
+
+bool test_set_and_get_on_device(sycl::queue q) {
+  sycl::buffer<int> int_buffer(1);
+  sycl::buffer<double> double_buffer(1);
+  sycl::buffer<custom_type> custom_type_buffer(1);
+
+  int new_int_value = 8;
+  double new_double_value = 3.0;
+  custom_type new_custom_type_value('b', 1.0, 12);
+
+  q.submit([&](sycl::handler &cgh) {
+    auto int_acc = int_buffer.get_access<sycl::access::mode::write>(cgh);
+    auto double_acc = double_buffer.get_access<sycl::access::mode::write>(cgh);
+    auto custom_type_acc =
+        custom_type_buffer.get_access<sycl::access::mode::write>(cgh);
+
+    cgh.set_specialization_constant<int_id>(new_int_value);
+    cgh.set_specialization_constant<double_id>(new_double_value);
+    cgh.set_specialization_constant<custom_type_id>(new_custom_type_value);
+
+    cgh.single_task<TestSetAndGetOnDevice>([=](sycl::kernel_handler kh) {
+      int_acc[0] = kh.get_specialization_constant<int_id>();
+      double_acc[0] = kh.get_specialization_constant<double_id>();
+      custom_type_acc[0] = kh.get_specialization_constant<custom_type_id>();
+    });
+  });
+
+  auto int_acc = int_buffer.get_access<sycl::access::mode::read>();
+  if (!check_value(new_int_value, int_acc[0],
+                   "integer specialization constant"))
+    return false;
+
+  auto double_acc = double_buffer.get_access<sycl::access::mode::read>();
+  if (!check_value(new_double_value, double_acc[0],
+                   "double specialization constant"))
+    return false;
+
+  auto custom_type_acc =
+      custom_type_buffer.get_access<sycl::access::mode::read>();
+  if (!check_value(new_custom_type_value, custom_type_acc[0],
+                   "custom_type specialization constant"))
+    return false;
+
+  return true;
+}
```

Comment:
For completeness, kernel_bundle-API based should also be tested. I guess these would be dozen extra lines of code. Can be addressed in a separate PR of course.

Reply:
> For completeness, kernel_bundle-API based should also be tested. I guess these would be dozen extra lines of code.

Sure, I definitively don't want to add those into the same test as it is huge enough already, but a separate file with such tests is needed for sure